packages/firestore/test/unit/api/database.test.ts

```diff
@@ -342,15 +486,173 @@ describe('QuerySnapshot', () => {
+        false,
+        false
       ).toJSON()
     ).to.throw(
       `QuerySnapshot.toJSON() attempted to serialize a document with pending writes. ` +
         `Await waitForPendingWrites() before invoking toJSON().`
     );
   });
+
+  it('fromJSON() throws with invalid data', () => {
+    const db = newTestFirestore();
+    expect(() => {
+      querySnapshotFromJSON(db, {});
+    }).to.throw;
+  });
+
+  it('fromJSON() throws with missing type data', () => {
+    const db = newTestFirestore();
+    expect(() => {
+      querySnapshotFromJSON(db, {
+        bundleSource: 'QuerySnapshot',
+        bundleName: 'test name',
+        bundle: 'test bundle'
+      });
+    }).to.throw;
+  });
+
+  it('fromJSON() throws with invalid type data', () => {
+    const db = newTestFirestore();
+    expect(() => {
+      querySnapshotFromJSON(db, {
+        type: 1,
+        bundleSource: 'QuerySnapshot',
+        bundleName: 'test name',
+        bundle: 'test bundle'
+      });
+    }).to.throw;
+  });
+
+  it('fromJSON() throws with invalid type data', () => {
+    const db = newTestFirestore();
+    expect(() => {
+      querySnapshotFromJSON(db, {
+        type: QuerySnapshot._jsonSchemaVersion,
+        bundleName: 'test name',
+        bundle: 'test bundle'
+      });
+    }).to.throw;
+  });
+
+  it('fromJSON() throws with invalid bundleSource type', () => {
+    const db = newTestFirestore();
+    expect(() => {
+      querySnapshotFromJSON(db, {
+        type: QuerySnapshot._jsonSchemaVersion,
+        bundleSource: 1,
+        bundleName: 'test name',
+        bundle: 'test bundle'
+      });
+    }).to.throw;
+  });
+
+  it('fromJSON() throws with invalid bundleSource value', () => {
+    const db = newTestFirestore();
+    expect(() => {
+      querySnapshotFromJSON(db, {
+        type: QuerySnapshot._jsonSchemaVersion,
+        bundleSource: 'DocumentSnapshot',
+        bundleName: 'test name',
+        bundle: 'test bundle'
+      });
+    }).to.throw;
+  });
+
+  it('fromJSON() throws with missing bundleName', () => {
+    const db = newTestFirestore();
+    expect(() => {
+      querySnapshotFromJSON(db, {
+        type: QuerySnapshot._jsonSchemaVersion,
+        bundleSource: 'QuerySnapshot',
+        bundle: 'test bundle'
+      });
+    }).to.throw;
+  });
+
+  it('fromJSON() throws with invalid bundleName', () => {
+    const db = newTestFirestore();
+    expect(() => {
+      querySnapshotFromJSON(db, {
+        type: QuerySnapshot._jsonSchemaVersion,
+        bundleSource: 'QuerySnapshot',
+        bundleName: 1,
+        bundle: 'test bundle'
+      });
+    }).to.throw;
+  });
+
+  it('fromJSON() throws with missing bundle data', () => {
+    const db = newTestFirestore();
+    expect(() => {
+      querySnapshotFromJSON(db, {
+        type: QuerySnapshot._jsonSchemaVersion,
+        bundleSource: 'QuerySnapshot',
+        bundleName: 'test name'
+      });
+    }).to.throw;
```

Comment:
IMO, it's usually good to assert that the error message is like expected. Mostly because bad or incorrect error messages can make for bad DX. Just a comment for your consideration.

Reply:
Done!